### PR TITLE
Alarm for excessive number of requests blocked by WAF (#5286)

### LINKED ISSUE #5286: Alarm for excessive number of requests blocked by WAF
Raise an alarm if a WAF blocks more than 10 requests per 5min.

- [x] Security design review completed; the Resolution of this issue does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] ~… affect monitoring of the system~
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: #5286

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] PR and connected issue are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by dsotirho-ucsc** [comment]
<details>

<summary>Terraform plan</summary>

```
Terraform will perform the following actions:

  # aws_cloudwatch_metric_alarm.waf will be created
  + resource "aws_cloudwatch_metric_alarm" "waf" {
      + actions_enabled                       = true
      + alarm_actions                         = [
          + "arn:aws:sns:us-east-1:122796619775:azul-monitoring-dev",
        ]
      + alarm_name                            = "azul-waf-daniel"
      + arn                                   = (known after apply)
      + comparison_operator                   = "GreaterThanThreshold"
      + dimensions                            = {
          + "Region" = "us-east-1"
          + "Rule"   = "ALL"
          + "WebACL" = "azul-api_gateway-daniel"
        }
      + evaluate_low_sample_count_percentiles = (known after apply)
      + evaluation_periods                    = 1
      + id                                    = (known after apply)
      + metric_name                           = "BlockedRequests"
      + namespace                             = "AWS/WAFV2"
      + ok_actions                            = [
          + "arn:aws:sns:us-east-1:122796619775:azul-monitoring-dev",
        ]
      + period                                = 300
      + statistic                             = "Sum"
      + tags                                  = {
          + "Name"       = "azul-waf-daniel"
          + "billing"    = "hca"
          + "component"  = "azul-waf"
          + "deployment" = "daniel"
          + "owner"      = "dsotirho@ucsc.edu"
          + "service"    = "azul"
        }
      + tags_all                              = {
          + "Name"       = "azul-waf-daniel"
          + "billing"    = "hca"
          + "component"  = "azul-waf"
          + "deployment" = "daniel"
          + "owner"      = "dsotirho@ucsc.edu"
          + "service"    = "azul"
        }
      + threshold                             = 1500
      + treat_missing_data                    = "notBreaching"
    }
```

</details>


## Blocked requests (sum over 5 minutes)

### Personal deployments:
waf-personal

### dev:
waf-dev

### anvildev:
waf-anvildev

### sandbox:
waf-sandbox

### anvilbox:
waf-anvilbox

### hammerbox:
waf-hammerbox

### prod:
(Image omitted for security purposes)
Max value: 5,192
Common spike: 1,000 - 3,000

### anvilprod:
(Image omitted for security purposes)
Max value: 8,928
Common spike: 500 - 1,000

**Review by achave11-ucsc — COMMENTED** [review]
Changes look good. 
However, I'd like for this changes to be tested in `anvildev` or `dev`. We should trigger a reindex, (it be nice to run the locus script also, to simulate, say 2 users and see some service traffic) to ensure we won't be bugged down with more alarms of this sort, of obscure provenance.

Also, does it seem strange all those high WAF blocked request for the lower deployments?

**Comment by dsotirho-ucsc** [comment]
The WAF blocked request counts have increased a bit in the last two weeks (see `anvildev` & `anvilbox`). I've updated the screenshots in [my previous post](https://github.com/DataBiosphere/azul/pull/5455#issuecomment-1670458279) and adjusted the thresholds set in this PR accordingly.

**Review by achave11-ucsc — CHANGES_REQUESTED** [review]
Changes look good, however upgrading instructions need to be carried out by developers and the PR CL is missing an item to remind everyone.

**Comment by dsotirho-ucsc** [comment]
> the PR CL is missing an item to remind everyone

Added a reminder CL item immediately after `Pushed merge commit to GitHub`

**Review by achave11-ucsc — DISMISSED** [review]

File: environment.py
Comment (hannes-ucsc) [inline environment.py]: Tying the threshold to a deployment is coincidental. Which deployment becomes victim to an attack is entirely out of our hands. It could hit production or a personal deployment. Ideally we'd make the metric a ratio between the numbers of served and blocked requests. If that's not possible we need to use a threshold that's "normal" for `prod`. In that case, please request a PL slot to discuss what that "norm" is.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by dsotirho-ucsc** [comment]
Note, checking this metric query on `prod` shows that this alarm would of fired a few times in the last 4 weeks.

Screen Shot 2023-09-14 at 2 05 37 AM

File: environment.py
Comment (hannes-ucsc) [inline environment.py]: Remove.

File: terraform/cloudwatch.tf.json.template.py
Comment (hannes-ucsc) [inline terraform/cloudwatch.tf.json.template.py]: Does this mean we'd be allowing 50% of the request to be blocked?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by dsotirho-ucsc** [comment]
> Does this mean we'd be allowing 50% of the request to be blocked?

Yes, the current settings would trigger the alarm when greater than 50% of requests are blocked, however it looks like this might cause the alarm to trigger quite frequently.

[Metric query % blocked on hcadev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-dev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

blocked-hca-dev

[Metric query % blocked on hcaprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1~stat~'Sum))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-prod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

blocked-hca-prod

[Metric query % blocked on anvildev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvildev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

blocked-anvil-dev

[Metric query % blocked on anvilprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100~label~'Expression1~id~'e1))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvilprod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~300~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d)

blocked-anvil-prod

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
>> Does this mean we'd be allowing 50% of the request to be blocked?

> Yes, the current settings would trigger the alarm when greater than 50% of requests are blocked, however it looks like this might cause the alarm to trigger quite frequently.

That wouldn't work, obviously. What's your plan on addressing this?

**Comment by dsotirho-ucsc** [comment]
Decided in parking lot discussion the alarm should trip when a threshold of 25% requests blocked is exceeded for 4 out of 4 evaluation periods, each period spanning 15 minutes.

With this metric query, the alarm would not have tripped in the last four weeks, however it would have come close with 3 consecutive 15 minute periods exceeding 25% blocked requests occurring twice on `hcaprod` (09/10 and 09/11) within the last 4 weeks, and once on `anvildev` (09/03).

[% blocked requests > 25% on hcadev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-dev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

blocked-hcadev

[% blocked requests > 25% on hcaprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~stat~'Sum~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-prod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

blocked-hcaprod

[% blocked requests > 25% on anvildev](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvildev~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

blocked-anvildev

[% blocked requests > 25% on anvilprod](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#metricsV2?graph=~(metrics~(~(~(expression~'m2*2f*28m1*2bm2*29*2a100*20*3e*2025~label~'Expression1~id~'e1~period~900))~(~'AWS*2fWAFV2~'AllowedRequests~'WebACL~'azul-api_gateway-anvilprod~'Region~'us-east-1~'Rule~'ALL~(id~'m1~visible~false))~(~'.~'BlockedRequests~'.~'.~'.~'.~'.~'.~(id~'m2~visible~false)))~view~'timeSeries~stacked~false~region~'us-east-1~stat~'Sum~period~900~start~'-PT672H~end~'P0D)&query=~'*7bAWS*2fWAFV2*2cRegion*2cRule*2cWebACL*7d):

blocked-anvilprod

**Comment by dsotirho-ucsc** [comment]
Creation of the alarm tested on my personal deployment.

Screen Shot 2023-09-20 at 6 12 52 PM

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**
